File: labs/aplicacao-exemplo/app-python/instrumentation.py
"""
⚡ INSTRUMENTAÇÃO OPEN TELEMETRY PARA SIGNOZ - PYTHON

Este módulo configura automaticamente a coleta de:
- Traces (rastreamento de requisições)
- Métricas (performance, contadores)
- Logs (eventos da aplicação)

Quando importado antes da aplicação Flask, ele:
1. Configura o SDK do OpenTelemetry
2. Habilita auto-instrumentação de bibliotecas populares
3. Exporta os dados para o SigNoz via Otel Collector
"""

import logging
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION, DEPLOYMENT_ENVIRONMENT
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.instrumentation.flask import FlaskInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor


def setup_instrumentation():
    """
    🔧 CONFIGURA O OPEN TELEMETRY SDK
    """
    
    # ----------------------------------------------------------------------------
    # RECURSO: Identifica sua aplicação
    # ----------------------------------------------------------------------------
    resource = Resource.create({
        SERVICE_NAME: "signoz-example-python",
        SERVICE_VERSION: "1.0.0",
        DEPLOYMENT_ENVIRONMENT: "development",
    })
    
    # ----------------------------------------------------------------------------
    # CONFIGURAÇÃO DE TRACES
    # ----------------------------------------------------------------------------
    
    # Provider de Traces: Gerencia e coleta traces
    tracer_provider = TracerProvider(resource=resource)
    
    # Exportador de Traces: Envia traces para o SigNoz
    otlp_trace_exporter = OTLPSpanExporter(
        endpoint="http://localhost:4317",  # Endpoint do Otel Collector
        insecure=True,  # Por padrão é HTTP, não HTTPS
    )
    
    # Processador de Spans: Agrupa spans em batches para envio eficiente
    span_processor = BatchSpanProcessor(otlp_trace_exporter)
    tracer_provider.add_span_processor(span_processor)
    
    # Ativa o provider de traces
    trace.set_tracer_provider(tracer_provider)
    
    # ----------------------------------------------------------------------------
    # CONFIGURAÇÃO DE MÉTRICAS
    # ----------------------------------------------------------------------------
    
    # Exportador de Métricas
    otlp_metric_exporter = OTLPMetricExporter(
        endpoint="http://localhost:4317",
        insecure=True,
    )
    
    # Leitor de Métricas: Exporta métricas periodicamente (a cada 60s)
    metric_reader = PeriodicExportingMetricReader(
        otlp_metric_exporter,
        export_interval_millis=60000,  # 60 segundos
    )
    
    # Provider de Métricas
    meter_provider = MeterProvider(
        resource=resource,
        metric_readers=[metric_reader],
    )
    
    # Nota: O SDK padrão do Python não expõe um global MeterProvider facilmente
    # Para métricas customizadas, você criaria assim:
    # from opentelemetry import metrics
    # metrics.set_meter_provider(meter_provider)
    
    # ----------------------------------------------------------------------------
    # AUTO-INSTRUMENTAÇÃO
    # ----------------------------------------------------------------------------
    
    # Flask: Instrumenta automaticamente todas as rotas
    FlaskInstrumentor().instrument()
    
    # Requests: Instrumenta chamadas HTTP externas
    RequestsInstrumentor().instrument()
    
    # ----------------------------------------------------------------------------
    # LOGGING
    # ----------------------------------------------------------------------------
    
    # Configura logs do OpenTelemetry
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
    )
    
    # ----------------------------------------------------------------------------
    # SUCESSO!
    # ----------------------------------------------------------------------------
    
    print("⚡ OpenTelemetry SDK inicializado")
    print("📊 Enviando traces e métricas para: http://localhost:4317")
    print("🔍 Dados aparecerão no SigNoz em: http://localhost:8080\n")


# ----------------------------------------------------------------------------
# IMPORTANTE: Execute a configuração quando o módulo é importado
# ----------------------------------------------------------------------------
setup_instrumentation()


"""
📚 CONCEITOS IMPORTANTES:

1. TRACE: Rastreia uma requisição HTTP única através do sistema
2. SPAN: Cada operação dentro de um trace (ex: chamada DB, API externa)
3. METRIC: Valores numéricos medidos ao longo do tempo
4. ATTRIBUTE: Metadados anexados a traces/spans
5. CONTEXT: Propaga informações através de diferentes serviços

🎯 O QUE VOCÊ GANHA:

- 🔍 Traces: Veja exatamente como cada requisição flui pela aplicação
- 📊 Métricas: Monitore performance, erros, throughput
- 🐛 Debug: Identifique gargalos e erros rapidamente
- 📈 Alertas: Configure alertas automáticos

🚀 PRÓXIMOS PASSOS:

1. Execute: pip install -r requirements.txt
2. Execute: python app.py
3. Acesse: http://localhost:8080 (SigNoz)
4. Explore os dados em tempo real!
"""

File: labs/aplicacao-exemplo/app-nodejs/instrumentation.js
/**
 * ⚡ INSTRUMENTAÇÃO OPEN TELEMETRY PARA SIGNOZ
 * 
 * Este arquivo configura automaticamente a coleta de:
 * - Traces (rastreamento de requisições)
 * - Métricas (performance, contadores)
 * - Logs (eventos da aplicação)
 * 
 * Quando você importa este arquivo com -r, ele:
 * 1. Configura o SDK do OpenTelemetry
 * 2. Habilita auto-instrumentação de bibliotecas populares
 * 3. Exporta os dados para o SigNoz via Otel Collector
 */

const { NodeSDK } = require('@opentelemetry/sdk-node');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-grpc');
const { Resource } = require('@opentelemetry/resources');
const { SemanticResourceAttributes } = require('@opentelemetry/semantic-conventions');

/**
 * 🔧 CONFIGURAÇÃO DO OPEN TELEMETRY SDK
 */
const sdk = new NodeSDK({
  /**
   * 📦 RECURSO: Identifica sua aplicação
   * 
   * Os atributos semânticos ajudam o SigNoz a:
   * - Identificar de qual serviço vêm os dados
   * - Filtrar e agrupar dados
   * - Mostrar informações úteis na UI
   */
  resource: new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: 'signoz-example-nodejs',
    [SemanticResourceAttributes.SERVICE_VERSION]: '1.0.0',
    [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: 'development',
  }),

  /**
   * 🔄 EXPORTADOR DE TRACES
   * 
   * Onde enviar os traces (rastreamento de requisições)
   * - OTLP: Open Telemetry Protocol
   * - gRPC: Protocolo de comunicação
   * - Endpoint: Otel Collector na porta 4317
   */
  traceExporter: new OTLPTraceExporter({
    url: 'http://localhost:4317', // Endpoint do Otel Collector
    // Opcional: adicionar headers de autenticação se necessário
    // headers: { 'Authorization': 'Bearer token' }
  }),

  /**
   * 📊 EXPORTADOR DE MÉTRICAS
   * 
   * Métricas são coletadas automaticamente pelo SDK
   * Não é necessário configurar um exportador separado para esta demo
   */

  /**
   * ⚡ AUTO-INSTRUMENTAÇÃO
   * 
   * Isso habilita automaticamente a instrumentação para:
   * ✅ HTTP/HTTPS requests
   * ✅ Express framework
   * ✅ PostgreSQL, MySQL, MongoDB
   * ✅ Redis
   * ✅ GraphQL
   * ✅ E muito mais!
   * 
   * Você não precisa modificar seu código manualmente!
   */
  instrumentations: [
    getNodeAutoInstrumentations({
      // Pode desabilitar instrumentações específicas se não usar
      '@opentelemetry/instrumentation-fs': {
        enabled: false, // Desabilita instrumentação de filesystem
      },
      // '@opentelemetry/instrumentation-express': {
      //   enabled: true,
      // },
    }),
  ],

  /**
   * 📝 LOGGER
   * 
   * Configura logs do OpenTelemetry (debug, info, warn, error)
   */
  // serviceName: 'signoz-example-nodejs', // Opcional, já definido no Resource
});

/**
 * ▶️ INICIALIZA O SDK
 * 
 * Esta linha ATIVA a instrumentação.
 * Tudo que acontecer após isso será automaticamente rastreado!
 */
sdk.start();
console.log('⚡ OpenTelemetry SDK inicializado');
console.log('📊 Enviando traces e métricas para: http://localhost:4317');
console.log('🔍 Dados aparecerão no SigNoz em: http://localhost:8080\n');

/**
 * 🛑 LIMPEZA AO ENCERRAR
 * 
 * Quando a aplicação é encerrada, desliga o SDK corretamente
 */
process.on('SIGTERM', () => {
  sdk.shutdown()
    .then(() => console.log('\n🔌 Telemetria encerrada'))
    .catch((error) => console.log('\n❌ Erro ao encerrar telemetria:', error))
    .finally(() => process.exit(0));
});

module.exports = sdk;

/**
 * 📚 CONCEITOS IMPORTANTES:
 * 
 * 1. TRACE: Rastreia uma requisição HTTP única através do sistema
 * 2. SPAN: Cada operação dentro de um trace (ex: chamada DB, API externa)
 * 3. METRIC: Valores numéricos medidos ao longo do tempo
 * 4. ATTRIBUTE: Metadados anexados a traces/spans
 * 5. CONTEXT: Propaga informações através de diferentes serviços
 * 
 * 🎯 O QUE VOCÊ GANHA:
 * 
 * - 🔍 Traces: Veja exatamente como cada requisição flui pela aplicação
 * - 📊 Métricas: Monitore performance, erros, throughput
 * - 🐛 Debug: Identifique gargalos e erros rapidamente
 * - 📈 Alertas: Configure alertas automáticos
 * 
 * 🚀 PRÓXIMOS PASSOS:
 * 
 * 1. Execute: npm install
 * 2. Execute: npm start
 * 3. Acesse: http://localhost:8080 (SigNoz)
 * 4. Explore os dados em tempo real!
 */

